fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/MacroFunction.java

```diff
@@ -0,0 +1,97 @@
+ */
+public class MacroFunction implements SerializableFunction {
+    @Nonnull
+    private final String functionName;
+
+    @Nonnull
+    private final List<Type> parameterTypes;
+    @Nonnull
+    private final Value bodyValue;
+    private final List<CorrelationIdentifier> parameterIdentifiers;
+
+    public MacroFunction(@Nonnull final String functionName, @Nonnull final List<QuantifiedObjectValue> parameters, @Nonnull final Value bodyValue) {
+        this.functionName = functionName;
+        this.parameterTypes = parameters.stream().map(QuantifiedObjectValue::getResultType).collect(Collectors.toUnmodifiableList());
+        this.parameterIdentifiers = parameters.stream().map(QuantifiedObjectValue::getAlias).collect(Collectors.toList());
+        this.bodyValue = bodyValue;
+    }
+
+    @Nonnull
+    public Value encapsulate(@Nonnull List<? extends Typed> arguments) {
+        // replace the QuantifiedObjectValue in body with arguments
+        SemanticException.check(arguments.size() == parameterTypes.size(), SemanticException.ErrorCode.FUNCTION_UNDEFINED_FOR_GIVEN_ARGUMENT_TYPES, "argument length doesn't match with function definition");
+        TranslationMap.Builder translationMapBuilder = new TranslationMap.Builder();
+        for (int i = 0; i < arguments.size(); i++) {
+            // check that arguments[i] type matches with parameterTypes[i]
+            final int finalI = i;
+            SemanticException.check(arguments.get(finalI).getResultType().equals(parameterTypes.get(i)), SemanticException.ErrorCode.FUNCTION_UNDEFINED_FOR_GIVEN_ARGUMENT_TYPES, "argument type doesn't match with function definition");
+            translationMapBuilder.when(parameterIdentifiers.get(finalI)).then((sourceAlias, leafValue) -> (Value)arguments.get(finalI));
+        }
+        return bodyValue.translateCorrelations(translationMapBuilder.build());
+    }
+
+    @Nonnull
+    @Override
+    public RecordMetaDataProto.SerializableFunction toProto() {
+        PlanSerializationContext serializationContext = new PlanSerializationContext(DefaultPlanSerializationRegistry.INSTANCE,
+                PlanHashable.CURRENT_FOR_CONTINUATION);
+        PMacroFunctionValue.Builder builder = PMacroFunctionValue.newBuilder();
+        for (int i = 0; i < parameterTypes.size(); i++) {
+            builder.addArguments(QuantifiedObjectValue.of(parameterIdentifiers.get(i), parameterTypes.get(i)).toValueProto(serializationContext));
+        }
+        return RecordMetaDataProto.SerializableFunction.newBuilder()
+                .setMacroFunction(builder
+                        .setFunctionName(functionName)
+                        .setBody(bodyValue.toValueProto(serializationContext)))
+                .build();
+
+    }
+
+    @Nonnull
+    public static MacroFunction fromProto(@Nonnull final PlanSerializationContext serializationContext, @Nonnull final PMacroFunctionValue functionValue) {
```

Comment:
I understand that I can inject serializationContext with @AutoService, but this method is called from `RecordMetaDataBuilder`, which doesn't have `serializationContext` injected. So it seems that I'll need to create a `serializationContext` in `RecordMetaDataBuilder` and pass it to `MacroFunction.fromProto` anyways.

Reply:
The context is normally used to do dynamic dispatch among the implementing classes.

Reply:
Got it. will do it in a follow-up PR.